src/dotnet-interactive-vscode/common/stdioKernelTransport.ts

```diff
@@ -160,45 +162,87 @@ export class StdioKernelTransport implements KernelTransport {
         });
     }
 
-    private handleLine(line: string) {
-        let obj = parse(line);
-        let envelope = <KernelEventEnvelope>obj;
-        switch (envelope.eventType) {
-            case DiagnosticLogEntryProducedType:
-                this.diagnosticChannel.appendLine((<DiagnosticLogEntryProduced>envelope.event).message);
-                break;
-        }
+    isKernelEventEnvelope(obj: any): obj is KernelEventEnvelope {
+        return obj.eventType
+            && obj.event;
+    }
+
+    isKernelCommandEnvelope(obj: any): obj is KernelCommandEnvelope {
+        return obj.commandType
+            && obj.command;
+    }
+
+    private async handleLine(line: string): Promise<void> {
+        const envelope = parse(line);
+        if (this.isKernelEventEnvelope(envelope)) {
+            switch (envelope.eventType) {
+                case DiagnosticLogEntryProducedType:
+                    this.diagnosticChannel.appendLine((<DiagnosticLogEntryProduced>envelope.event).message);
+                    break;
+            }
+
+            for (let i = this.eventSubscribers.length - 1; i >= 0; i--) {
+                this.eventSubscribers[i](envelope);
+            }
+        } else if (this.isKernelCommandEnvelope(envelope)) {
+            for (let i = this.commandSubscribers.length - 1; i >= 0; i--) {
+                // TODO: pass in context with shortcut methods for publish, etc.
+                // TODO: wrap and return succeed/failed
+                try {
+                    const handled = await this.commandSubscribers[i](envelope);
+                    if (handled) {
```

Comment:
This becomes dependent on the subscription order when there are multiple subscribers that might be added in a different order. We should enforce a exactly one command handler.

Reply:
I think here we ensure a single subscriber for commands, that would actually be the kernel handling them , same job we do in the .NET World using a composite kernel. here we need a kernel with routing responsibility too.

Reply:
Updated to only have a single command handler instead of a list.